src/main/java/fi/vm/yti/datamodel/api/v2/service/SchemaService.java

```diff
@@ -1,161 +1,264 @@
+	/**
+	 * 
+	 * Adds an object property to the RDF model.
+	 * 
+	 * @param propID      The property ID.
+	 * @param node        The JSON node containing the property details.
+	 * @param model       The RDF model.
+	 * @param schemaPID   The schema PID.
+	 * @param targetShape The target shape for the object property.
+	 * @return The created resource representing the object property.
+	 */
+	private Resource addObjectProperty(String propID, JsonNode node, Model model, String schemaPID,
+			String targetShape) {
 		Resource propertyResource = model.createResource(schemaPID + "#" + propID);
-		propertyResource.addProperty(RDF.type, model.getResource(SHACL.PropertyShape.getURI()));
+		propertyResource.addProperty(RDF.type, model.getResource(SH.PropertyShape.getURI()));
 		propertyResource.addProperty(DCTerms.type, OWL.ObjectProperty);
-		if(node.get("type").asText().equals("array")) {
-			if(node.has("maxItems"))
-				propertyResource.addProperty(model.getProperty(SHACL.maxCount.getURI()), model.createTypedLiteral(node.get("maxItems").asInt()));
-			if(node.has("minItems"))
-				propertyResource.addProperty(model.getProperty(SHACL.minCount.getURI()), model.createTypedLiteral(node.get("minItems").asInt()));
 
-		}
-		else {
-			propertyResource.addProperty(model.getProperty(SHACL.maxCount.getURI()), model.createTypedLiteral(1));
-			propertyResource.addProperty(model.getProperty(SHACL.minCount.getURI()), model.createTypedLiteral(1));
-			
-		}
-		propertyResource.addProperty(model.getProperty(SHACL.name.getURI()), propID);
-		propertyResource.addProperty(model.getProperty(SHACL.path.getURI()), propID);
-		propertyResource.addProperty(model.getProperty(SHACL.node.getURI()), model.createResource(targetShape));
-				
+		checkAndAddPropertyFeature(node, model, propertyResource);
+		propertyResource.addProperty(model.getProperty(SH.path.getURI()), propID);
+		propertyResource.addProperty(model.getProperty(SH.node.getURI()), model.createResource(targetShape));
+
 		return propertyResource;
-		
-	}		
-	
+	}
 
-	/**
+	private void handleDatatypeProperty(String propID, Entry<String, JsonNode> entry, Model model, String schemaPID,
+			Resource nodeShapeResource, boolean isRequired, boolean isArrayItem) {
+		Resource propertyResource = addDatatypeProperty(propID + "/" + entry.getKey(), entry.getValue(), model,
+				schemaPID, entry.getValue().get("type").asText());
+		nodeShapeResource.addProperty(model.getProperty(SH.property.getURI()), propertyResource);
+		if (!isArrayItem) {
+			handleRequiredProperty(entry.getValue(), model, propertyResource, isRequired);
+		} 
+		if (entry.getValue().get("type").asText().equals("string") & entry.getValue().has("pattern")) {
+			propertyResource.addProperty(model.getProperty(SH.pattern.getURI()), entry.getValue().get("pattern").asText());
+		}
+	}
 	
-	Handles an object property and creates the corresponding SHACL (Node)Shape.
+	private String capitaliseNodeIdentifier(String propID) {
```

Comment:
Considered optimizing this. Would lastInfoOf be enough to get the index to capitalize without the need to create a LinkedList?